test/tests/functional/pbs_offline_vnodes.py

```diff
@@ -0,0 +1,262 @@
+            self.is_cray = True
+
+        TestFunctional.setUp(self)
+        self.server.manager(MGR_CMD_SET, SERVER, {'log_events': 2047})
+
+    def create_mom_hook(self):
+        name = "h1"
+        body = ("import pbs\n"
+                "import time\n"
+                "time.sleep(60)\n"
+                "pbs.event.accept()")
+        attr = {'event': 'execjob_begin', 'fail_action': 'offline_vnodes',
+                'alarm': '3', 'enabled': 'true'}
+        self.server.create_import_hook(name, attr, body)
+
+    def create_multi_vnodes(self, num_moms):
+        if self.is_cray is True:
+            return
+        vn_attrs = {ATTR_rescavail + '.ncpus': 1,
+                    ATTR_rescavail + '.mem': '1024mb'}
+        for i in range(num_moms):
+            self.server.create_vnodes('vnode', vn_attrs, 3, self.moms.values()[
+                                      i], usenatvnode=True, delall=False,
+                                      expect=False)
+
+    def test_single_mom_hook_failure_affects_vnode(self):
+        """
+        Run an execjob_begin hook that sleep for sometime,
+        at the same time set an alarm value so less that
+        the hook alarms out and server executes the fail_action
+        After this check if vnodes are marked offline.
+        In case of a single mom reporting vnodes, it should mark
+        all the vnodes and mom as offline.
+        Once offlined, reset the mom by issueing pbsnodes -r
+        and check if the job runs on one of the vnodes.
+        """
+        single_mom = ""
+        if len(self.moms.values()) != 1:
+            self.server.manager(MGR_CMD_DELETE, NODE, id="@default",
+                                expect=True)
+        single_mom = self.moms.values()[0]
+        # since we deleted all nodes, we need to create one for cray
+        # because create_multi_vnodes does not create vnode on cray.
+        if self.is_cray is True:
+            self.server.manager(MGR_CMD_CREATE, NODE,
+                                id=single_mom.shortname)
+        # adding a sleep of two seconds because it takes some time before
+        # node resources start showing up
+            time.sleep(2)
+        # No need to create vnodes on a cpuset mom
+        cpuset_mom = single_mom.is_cpuset_mom()
+        if cpuset_mom is False:
+            self.create_multi_vnodes(1)
+        self.create_mom_hook()
+
+        j1 = Job(TEST_USER)
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
+                           id=single_mom.shortname, max_attempts=3,
+                           interval=2)
+        j1.set_sleep_time(1000)
+        jid = self.server.submit(j1)
+        self.server.expect(JOB, {ATTR_state: 'Q'}, id=jid)
+
+        self.server.expect(
+            NODE, {ATTR_NODE_state: 'offline'},
+            id=single_mom.shortname, max_attempts=20, interval=2)
+
+        vname = "vnode[0]"
+        if self.is_cray is True:
+            vlist = []
+            first_vnode = 1

```

Comment:
how about setting vname to None when you enter the if block and just checking for vname inside the "for v1 in vlist" loop? if it's None, set it to v1. That way you won't need 'first_vnode' at all.

Reply:
in fact, you don't really need to set vname before the if-else blocks, you could just check whether vname is None after the if-else blocks, if it is then set it to 'vnode[0]'